static/app/components/globalModal.tsx

```diff
@@ -1,97 +1,276 @@
+   */
+  visible: boolean;
+  /**
+   * A function that returns a React Element
    */
   children?: null | ((renderProps: ModalRenderProps) => React.ReactNode);
-
   /**
-   * Note this is the callback for the main App container and
-   * NOT the calling component.  GlobalModal is never used directly,
-   * but is controlled via stores. To access the onClose callback from
-   * the component, you must specify it when using the action creator.
+   * Note this is the callback for the main App container and NOT the calling
+   * component. GlobalModal is never used directly, but is controlled via
+   * stores. To access the onClose callback from the component, you must
+   * specify it when using the action creator.
    */
   onClose?: () => void;
 };
 
-class GlobalModal extends React.Component<Props> {
-  static defaultProps: DefaultProps = {
-    visible: false,
-    options: {},
-  };
+const BaseHeader = styled('header')`
+  position: relative;
+  border-bottom: 1px solid ${p => p.theme.border};
+  padding: ${space(3)} ${space(4)};
+  margin: -${space(4)} -${space(4)} ${space(3)} -${space(4)};
 
-  handleCloseModal = () => {
-    const {options, onClose} = this.props;
+  h1,
+  h2,
+  h3,
+  h4,
+  h5,
+  h6 {
+    font-size: 20px;
+    font-weight: 600;
+    margin-bottom: 0;
+    line-height: 1.1;
+  }
+`;
 
-    // onClose callback for calling component
-    if (typeof options.onClose === 'function') {
-      options.onClose();
-    }
+const Body = styled('div')`
+  font-size: 15px;
+
+  p:last-child {
+    margin-bottom: 0;
+  }
 
-    // Action creator
-    closeModal();
+  img {
+    max-width: 100%;
+  }
+`;
+
+const Footer = styled('footer')`
+  border-top: 1px solid ${p => p.theme.border};
+  display: flex;
+  justify-content: flex-end;
+  padding: ${space(3)} ${space(4)};
+  margin: ${space(3)} -${space(4)} -${space(4)};
+`;
+
+type HeaderProps = {
+  /**
+   * Show a close button in the header
+   */
+  closeButton?: boolean;
+};
+
+const makeClosableHeader = (closeModal: () => void) => {
+  const ClosableHeader: React.FC<
+    React.ComponentProps<typeof BaseHeader> & HeaderProps
+  > = ({closeButton, children, ...props}) => (
+    <BaseHeader {...props}>
+      {children}
+      {closeButton && <CloseButton onClick={closeModal} />}
+    </BaseHeader>
+  );
+
+  ClosableHeader.displayName = 'Header';
+
+  return ClosableHeader;
+};
+
+type ModalRenderProps = {
+  closeModal: () => void;
+  Header: ReturnType<typeof makeClosableHeader>;
+  Body: typeof Body;
+  Footer: typeof Footer;
+};
+
+const getModalPortal = memoize(() => {
+  let portal = document.getElementById('modal-portal');
+  if (!portal) {
+    portal = document.createElement('div');
+    portal.setAttribute('id', 'modal-portal');
+    document.body.appendChild(portal);
+  }
 
-    if (typeof onClose === 'function') {
-      onClose();
+  return portal as HTMLDivElement;
+});
+
+function GlobalModal({visible = false, options = {}, children, onClose}: Props) {
```

Comment:
should we move some of these components to a new file? maybe that way it would be easier to read

Reply:
Yeah I was thinking that too

Reply:
Moved!